ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java

```diff
@@ -0,0 +1,412 @@
+        monitor.leave();
+      }
+    }
+
+    return null;
+  }
+
+  private PullQueryRow pollRow() {
+    final HandledRow polled;
+    monitor.enter();
+    try {
+      polled = queue.poll();
+    } finally {
+      monitor.leave();
+    }
+
+    if (polled == null) {
+      return null;
+    }
+
+    polled.handler.handle(FUTURE_FACTORY.succeededFuture());
+
+    monitor.enter();
+    try {
+      if (queue.size() < softLimit) {
+        drainHandler.forEach(h -> h.handle(null));
+      }
+    } finally {
+      monitor.leave();
+    }
+
+    return polled.row;
+  }
+
+  public int getTotalRowsQueued() {
+    monitor.enter();
+    try {
+      // pass by value, this should be safe not to be changed elsewhere
+      return totalRowsQueued;
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  @Override
+  public int size() {
+    monitor.enter();
+    try {
+      return queue.size();
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return size() == 0;
+  }
+
+  public boolean isDone() {
+    monitor.enter();
+    try {
+      return closed || hardLimitHit();
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  private boolean hardLimitHit() {
+    monitor.enter();
+    try {
+      return hardLimit.isPresent() && totalRowsQueued >= hardLimit.getAsInt();
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  // -------------------- HANDLER SETTERS -------------------------
+
+
+  @Override
+  public PullQueryWriteStream exceptionHandler(final Handler<Throwable> handler) {
+    // we don't currently use this but may have use for it in the future
+    return this;
+  }
+
+  @Override
+  public PullQueryWriteStream drainHandler(final Handler<Void> handler) {
+    drainHandler.add(handler);
+    return this;
+  }
+
+  @Override
+  public void setCompletionHandler(final CompletionHandler handler) {
+    this.endHandler = handler;
+  }
+
+  @Override
+  public void setLimitHandler(final LimitHandler handler) {
+    this.limitHandler = ar -> handler.limitReached();
+  }
+
+  @Override
+  public void setQueuedCallback(final Runnable callback) {
+    final Runnable parent = queueCallback;
+
+    queueCallback = () -> {
+      parent.run();
+      callback.run();
+    };
+  }
+
+
+  // -------------------- WRITE STREAM METHODS -------------------------
+
+  public void putConsistencyVector(final ConsistencyOffsetVector offsetVector) {
+    write(ImmutableList.of(
+        StreamedRow.consistencyToken(new ConsistencyToken(offsetVector.serialize()))
+    ));
+  }
+
+  @Override
+  public PullQueryWriteStream write(final List<StreamedRow> data) {
+    write(data, ar -> { });
+    return this;
+  }
+
+  @Override
+  public PullQueryWriteStream write(
+      final List<StreamedRow> data,
+      final Handler<AsyncResult<Void>> handler
+  ) {
+    if (isDone()) {
+      return this;
+    }
+
+    monitor.enter();
+    try {
+      translator.add(data);
+      while (translator.hasNext()) {
```

Comment:
The translator can return `false` from `hasNext` when it has no more rows queued.  

What happens then?

Reply:
not sure I understand the question, that's the expected behavior?

Reply:
Sorry, I was still grokking things... I was expecting that `translator.add(data);` was being called on another thread.:)